packages/contracts/contracts/identity/IdentityProxy.s

```diff
@@ -0,0 +1,200 @@
+    function executeCreate(bytes memory data)
+        internal
+        returns (address newContract)
+    {
+        // solium-disable-next-line security/no-inline-assembly
+        assembly {
+            newContract := create(0, add(data, 0x20), mload(data))
+        }
+    }
+
+    function getHash(address signer, address to, uint value, bytes data)
+        public view
+        returns(bytes32)
+    {
+        return keccak256(abi.encodePacked(signer, to, value, data, nonce[signer]));
+    }
+
+    // original forward function copied from https://github.com/uport-project/uport-identity/blob/develop/contracts/Proxy.sol
+    function forward(address to, bytes sign, address signer, bytes data) public {
+        uint value = 0 ether;
+
+        // the hash contains all of the information about the meta transaction to be called
+        bytes32 _hash = getHash(signer, to, value, data);
+
+        // increment the hash so this tx can't run again
+        nonce[signer]++;
+
+        // this makes sure signer signed correctly AND signer is a valid bouncer
+        require(isSignedByOwner(_hash, sign), "Signer is not contract owner");
+
+        // execute the call
+        require(executeCall(to, value, data), "Cannot execute the call");
+        emit Forwarded(sign, signer, to, value, data, _hash);
+    }
+
+    // borrowed from OpenZeppelin's ESDA stuff:
+    // https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/cryptography/ECDSA.sol
+    function isSignedByOwner(bytes32 _hash, bytes _signature)
+        internal view
+        returns (bool)
+    {
+        bytes32 r;
+        bytes32 s;
+        uint8 v;
+
+        // Check the signature length
+        if (_signature.length != 65) {
+            return false;
+        }
+
+        // Divide the signature in r, s and v variables
+        // ecrecover takes the signature parameters, and the only way to get them
+        // currently is to use assembly.
+        // solium-disable-next-line security/no-inline-assembly
+        assembly {
+            r := mload(add(_signature, 32))
+            s := mload(add(_signature, 64))
+            v := byte(0, mload(add(_signature, 96)))
+        }
+
+        // Version of signature should be 27 or 28, but 0 and 1 are also possible versions
+        if (v < 27) {
+            v += 27;
+        }
+
+        // If the version is correct return the signer address
+        if (v != 27 && v != 28) {
+            return false;
+        } else {
+            // solium-disable-next-line arg-overflow
+            return owner == ecrecover(keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", _hash)), v, r, s);
+        }
+    }
+
+    function swapAndMakeOffer(
+        address _owner,
+        address _marketplace,
+        bytes _offer,
+        address _exchange,
+        bytes _swap,
+        address _token,
+        uint _value
+    )
+        public
+        payable
+    {
+        owner = _owner;
```

Comment:
Why override owner? Are we going to deploy a single proxy contract and make all users use that?

This function doesn't have `onlyOwner` modifier. Won't this allow any person to exploit this and take control of the contract?

Reply:
An attacker can use this to make himself an owner of the contract and change the `Marketplace` address to his own contract that transfers the tokens (deposited for creating listing after the address change) to his account.

Reply:
Ignore the previous comment. I thought that `_marketplace` is global variable. Didn't see that it was a function parameter. Anyway I'm still skeptical about the reassignment of owner address.